## What is autograd?

Autograd is one of the most important packages in Pytorch. The autograd package allows us to differentiate all tensors given an operation. Differentiation is a constituent of  backpropagation, which allows us to update the weights in a linear regression or in a Deep Neural Network. Thus autograd allows to smoothly perform backpropagation independent of the operations involved in-between. Let us delve into further details. 
Key features :
1. It is a define by the run framework, which implies that the backpropagation is defined by the code, and every single run can be different.

### The autograd variable

The autograd.Variable is the central class of the package. The Variable wraps around tensors and supports nearly all of pytorch tensor operators. Like in TensorFlow one also needs to initialize the variables. 

In [1]:
import torch

In [2]:
from torch.autograd import Variable

In [31]:
#Creating a variable
x = Variable(torch.ones(2, 2), requires_grad = True)

In [32]:
print(x)

Variable containing:
 1  1
 1  1
[torch.FloatTensor of size 2x2]



We can now perform various operation on the variable.

In [33]:
y = x + 2

In [34]:
print(y)

Variable containing:
 3  3
 3  3
[torch.FloatTensor of size 2x2]



The fact that the y was created as a result of an operation, leads to y acquiring an additional attribute knows as .grad_fn. 'y' has been created due to a particular function and thus it has the attribute. In other words, Variable and Function are actually interconnected. You will use functions to create variables and the created a variable will have attribute of the functions. Howver the variables directly created by the user do not have the function attribute as they have been the result of any pytorch operation. For the example in the above example the variable y has been created as a result of and addition operation and will thus have a grad_fn attribute

In [35]:
print(y.grad_fn)

In [36]:
#let's make more functions
z = 3*y*y + 4*y
out = z.mean()
print(z, out)

Variable containing:
 39  39
 39  39
[torch.FloatTensor of size 2x2]
 Variable containing:
 39
[torch.FloatTensor of size 1]



In [37]:
#Perform backpropagation
out.backward()

In [38]:
#Print t
print(x.grad)

Variable containing:
 5.5000  5.5000
 5.5000  5.5000
[torch.FloatTensor of size 2x2]



Let us check whether the output is correct or not. We are calculating the gradient $\frac{d out}{dx}$ at $x = 1$. Now $$out =\frac{1}{4}\sum z_{i}$$ where $ z_{i} = 3\left(x + 2\right)^2 + 4\left(x + 2\right) $. Now $$\frac{d\text{ }out}{dx} = \frac{3}{2}\left(x + 2\right) + 1 $$ We should get a $2\times 2$ tensor with each element having a value 5.5


In the next tutorial I will discuss how to proceed to make a simple neural network with the help of all the tools at hand.